torchaudio-contrib/io.py

```diff
@@ -0,0 +1,48 @@
+import torch
+import numpy as np
+import warnings
+from scipy.io import wavfile
+import librosa
+
+def load(filename, sr=44100, mono=True, norm=False, device='cpu', dtype=np.float32):
+    """
+    Generic wrapper for reading an audio file.
+    Different libraries offer different speeds for this, so this routine is the
+    'catch-all' for whatever read routine happens to work best
+
+    Tries a fast method via scipy first, reverts to slower librosa when necessary.
+    """
+    # first try to read via scipy, because it's fast
+    scipy_ok = False
+    with warnings.catch_warnings(record=True) as w:
+        warnings.simplefilter("error")    # scipy throws warnings which should be errors
+        try:
+            out_sr, signal = wavfile.read(filename)
+            scipy_ok = True
+        except wavfile.WavFileWarning:
+            print("read_audio_file: Warning raised by scipy. ",end="")
+
+    if scipy_ok:
+        if mono and (len(signal.shape) > 1):     # convert to mono (by truncating other channels)
+            signal = signal[:,0]
+
+        if isinstance(signal[0], np.int16):      # convert from ints to floats if necessary
+            signal = np.array(signal, dtype=dtype)/32767.0   # change from [-32767..32767] to [-1..1]
```

Comment:
This creates a copy of the full signal as the given dtype, even if the given dtype is `int16`, and then creates another copy of the full signal with the scaled values. Should be done in a single step with `np.multiply`.
Note that if you pass `np.int16` as the dtype, you will get a `float64` result (that will be converted back to int16 in the end, but takes up 4 times the space in the meantime).

Reply:
Thanks for catching this.   I was unaware that astype copied by default;  one can easily add `copy=False`, but yea, that dtype should go on the outside, not the inside!

Reply:
If the dtype goes on the outside (I assume you meant `np.array(signal / magic_number, dtype=dtype)`), you're still creating a temporary float64 copy that's converted afterwards. I'd suggest doing the division and conversion in one step using `np.multiply`.